Allow users to give feedback when they encounter an error page #4315

HirdayGupta opened the pull request:
Fixes #4315 

Whew. My first semi-large PR. 😅

**Outline of solution**

- Added some introductory text to `errorPage.tag` (generic error page for all types)
- Added a new tag, `errorPageEmailCompose.tag` which contains the feedback form to all error pages (GIF attached). The recipient email address field is non-editable but displayed (for the User's information) and is prefilled to match the value in `Config.SUPPORT_EMAIL`. This tag is included in `errorPage.tag`
- The form sends an AJAX POST request when submitted to the newly created `feedbackResponseEmailSendAction`.
- This action verifies if the user is logged in (I figured that this feature should be made available only to TEAMMATES users, lmk if you would like this changed) and then logs the message sent by the user (and user's account details) at the `severe` level. This is based on @damithc's suggestion as he mentioned that all logs at severe level get sent to as an email to admin.  [(#4315 comment)](https://github.com/TEAMMATES/teammates/issues/4315#issuecomment-308364391).
- The action then returns the AJAX status depending on whether the user was logged in or not. (Note: have made a base class `PageData` object in the action as I did not need any custom data).
- On return, the form disappears and the status message is shown using existing mechanism imported from `statusMessage.es6`.

**Possible Contentions/Advice needed from Senior Devs:**
- What should be the content of the success/error messages after submitting the form? (These are stored in `Const.StatusMessages.ERROR_FEEDBACK_SUBMIT_SUCCESS` and ``Const.StatusMessages.ERROR_FEEDBACK_SUBMIT_FAILED` respectively).
- Should I be adding the tinyMCE Rich Text Editor to the feedback content textarea in this form? Will this really be necessary?
- I figured it might be useful to have a pre-defined subject to distinguish these log emails from any other log emails sent to admin. Currently, it just reads `"Pre-defined subject"`. Any suggestions on what this should read and whether the user should be able to edit the field?
- For some reason the `try-catch` block surrounding the `gateKeeper.verifyLoggedInUserPrivileges()` in the `FeedbackResponseEmailSendAction` is not behaving as expected and is not catching the error thrown. As a result, the FAILED status message is never shown. (PS: I recognize that in the current implementation, I should enclose the `log.severe()` call in an `if(isUserLoggedOn)` but right now, the `catch` block is only not catching the error so I thought we could address that first).
- I was wondering if this feature would be more useful if the `severe log message` included details about the type of error. To implement this, I would add a Session Attribute in the part of the `ControllerServlet` file that calls the error pages like `req.getSession().setAttribute("ERROR_TYPE", <ERROR-TYPE>);` and then access this attribute in the DOM by setting the value of a hidden input like `{session.getAttribute("ERROR_TYPE")}`. The other way would be to define a data attribute in each error page's JSP file. I was wondering if this is needed and if yes, which approach would be better?
- Anything else you would like me to include the `severe log` email that might make it easier for the admin to pinpoint the error?

**GIF + Screenshot of new behavior:**

**GIF of form:**
![ezgif com-video-to-gif 2](https://user-images.githubusercontent.com/18088721/27809426-0fbba42a-606d-11e7-8c70-bc49722fce55.gif)


**Screenshot of Log after submitting form:**
![4315-log-output](https://user-images.githubusercontent.com/18088721/27809380-bf228baa-606c-11e7-837b-924ee9ccb0c5.png)

**[Link to full log output](https://gist.github.com/HirdayGupta/7c31e186818cb9df7719802f218b212a)**

**Link to staging server:** https://4-45-dot-teammates-hirday.appspot.com

**[Sample instructor join link for above server](https://4-45-dot-teammates-hirday.appspot.com/page/instructorCourseJoin?key=5BEA512C245AF0C78922FF5B29BE4B17B23BD4711A82A03CE7718C88E12ED7584B03CA78285FA71CE445594F44324AD9&instructorinstitution=NUS)**

damithc commented:
> What should be the content of the success/error messages after submitting the form? (These are stored in Const.StatusMessages.ERROR_FEEDBACK_SUBMIT_SUCCESS and ``Const.StatusMessages.ERROR_FEEDBACK_SUBMIT_FAILED` respectively).

Something like this?
`Your error report has been recorded. We will follow up with you in due course, usually, within 24 hours`
`Failed to record the error message. Please email our support team at teammates@comp.nus.edu.sg`

> Should I be adding the tinyMCE Rich Text Editor to the feedback content textarea in this form? Will this really be necessary?

I guess no need. The log cannot record rich text anyway.

> I figured it might be useful to have a pre-defined subject to distinguish these log emails from any other log emails sent to admin. Currently, it just reads "Pre-defined subject". Any suggestions on what this should read and whether the user should be able to edit the field?

`User-submitted Error Report`?

> For some reason the try-catch block surrounding the gateKeeper.verifyLoggedInUserPrivileges() in the FeedbackResponseEmailSendAction is not behaving as expected and is not catching the error thrown. As a result, the FAILED status message is never shown. (PS: I recognize that in the current implementation, I should enclose the log.severe() call in an if(isUserLoggedOn) but right now, the catch block is only not catching the error so I thought we could address that first).

Someone else need to answer this question. I haven't worked with that part of the code recently.

> I was wondering if this feature would be more useful if the severe log message included details about the type of error. To implement this, I would add a Session Attribute in the part of the ControllerServlet file that calls the error pages like req.getSession().setAttribute("ERROR_TYPE", <ERROR-TYPE>); and then access this attribute in the DOM by setting the value of a hidden input like {session.getAttribute("ERROR_TYPE")}. The other way would be to define a data attribute in each error page's JSP file. I was wondering if this is needed and if yes, which approach would be better?

Can be done as a separate PR.

> Anything else you would like me to include the severe log email that might make it easier for the admin to pinpoint the error?

Can be done as separate PRs.

HirdayGupta commented:
Hi @damithc. Thanks for answering the questions!

Have changed the SUCCESS status message and the pre-defined subject.

Just one more thing:

- The error message you suggested was:
    >Failed to record the error message. Please email our support team at teammates@comp.nus.edu.sg
    
    I forgot to mention that the only case in which this will be displayed is if the user tries to submit feedback and is not logged in. In light of this, would you recommend changing the message?

damithc commented:
> I forgot to mention that the only case in which this will be displayed is if the user tries to submit feedback and is not logged in. In light of this, would you recommend changing the message?

I think the suggested message still fits. No need to be very specific.

HirdayGupta commented:
@damithc Done. Success message, Error message and subject edited.

The only issue (which needs to be addressed by some other dev) is this:
>For some reason the try-catch block surrounding the gateKeeper.verifyLoggedInUserPrivileges() in the FeedbackResponseEmailSendAction is not behaving as expected and is not catching the error thrown. As a result, the FAILED status message is never shown. (PS: I recognize that in the current implementation, I should enclose the log.severe() call in an if(isUserLoggedOn) but right now, the catch block is only not catching the error so I thought we could address that first).

Other than that, this PR is **Ready for review.**

chowyb commented:
@HirdayGupta just a brief comment on why the catch block is currently not working: if a user is not logged in, the code sends a redirect to the login page in `ControllerServlet.doPost()`, instead of running the associated action's `execute()` method. As such, the reason why the catch block appeared to have done nothing was because execution did not even reach the try block.

Also, regarding naming of files, I would suggest a different name than `FeedbackResponse...`, as this may easily be confused with the existing kind of feedback response (ie. feedback questions/responses).

HirdayGupta commented:
> just a brief comment on why the catch block is currently not working: if a user is not logged in, the code sends a redirect to the login page in ControllerServlet.doPost(), instead of running the associated action's execute() method. As such, the reason why the catch block appeared to have done nothing was because execution did not even reach the try block.

@chowyb Any suggestions on how I can avoid this? Or is redirecting to the login page an acceptable behavior? If it is the latter, any suggestions on how I can make the `submit` button successfully redirect to the login page? It just does nothing at this point.

Regarding your second suggestion would `ErrorUserReportLogAction` be an acceptable file name?

chowyb commented:
> Any suggestions on how I can avoid this? Or is redirecting to the login page an acceptable behavior? If it is the latter, any suggestions on how I can make the submit button successfully redirect to the login page? It just does nothing at this point.

@HirdayGupta Redirecting to the login page might be acceptable behaviour, going by how non-AJAX calls behave (redirecting to the login page), however in the case of AJAX calls, this may be difficult since it returns a HTML page instead of a JSON object if the user is not logged in. From existing AJAX calls (using "remind all students" as an example), if the user is not logged in, then it appears to never return anything. Perhaps we could explore a way to handle this in a more global manner in another PR? For now, using the AJAX gif to show progress should suffice.

> Regarding your second suggestion would ErrorUserReportLogAction be an acceptable file name?

Yep, seems good.

HirdayGupta commented:
Hi @chowyb. 

- Renamed the file in https://github.com/TEAMMATES/teammates/pull/7675/commits/fee0c49653e9d280acbf00a73545d3874ddf9d4c.

- Added the AJAX spinner and siabled `submit feedback` button in the`beforeSend()` AJAX function.

> From existing AJAX calls (using "remind all students" as an example), if the user is not logged in, then it appears to never return anything. Perhaps we could explore a way to handle this in a more global manner in another PR? For now, using the AJAX gif to show progress should suffice.

- I managed to fix the "never returning anything" part of the error (only for this PR, global handling would definitely be a lot better) in the JS Ajax call, by checking if the `result` data returned by the ajax call was `undefined` (see line 40 from  6a302cb). If it is, that means nothing was returned. And here, instead of just showing the AJAX Gif (which seemed flawed as then the user would assume there was an error with the error reporting page! :p), I show the error status message (see GIF below). 

- Also, I've added an `error()` function in the ajax function which displays the same status message.

**GIF of new behaviour on error:**

![ezgif com-video-to-gif-2](https://user-images.githubusercontent.com/18088721/28086251-fe6b19f0-669b-11e7-835e-6f08267bf65d.gif)

chowyb commented on src/main/webapp/WEB-INF/tags/errorPageEmailCompose.tag:
```diff
@@ -0,0 +1,21 @@
+    <div class="form-group">
+        <label for="error-feedback-email-composer-recipient-email">To:</label>
+        <br><em>Note: This is the TEAMMATES support email. It has been pre-filled for your convenience and is not editable. </em>
+        <input type="email" value=<%= Config.SUPPORT_EMAIL %> id="error-feedback-email-composer-recipient-email" class="form-control" name=<%= Const.ParamsNames.ERROR_FEEDBACK_EMAIL_RECEIVER_ADDRESS %> readonly>
```
Use `readonly=""`

chowyb commented on src/main/webapp/dev/js/main/errorPageEmailComposer.es6:
```diff
@@ -0,0 +1,54 @@
+} from '../common/const.es6';
+
+const ERROR_STATUS_MESSAGE = 'Failed to record the error message. Please email our support team at '
+        + '<a href="mailto:teammates@comp.nus.edu.sg">teammates@comp.nus.edu.sg</a>.';
```
It might be better to put a constant for the email address in `const.es6`, and to use [template literals](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Template_literals) for this string.

HirdayGupta replied:
@chowyb Or should I take the value from the DOM element that draws it from `Config.SUPPORT_EMAIL`?

HirdayGupta replied:
Would implement it like this:
```JavaScript
const SUPPORT_EMAIL = $('#error-feedback-email-composer-recipient-email').prop('value');
const ERROR_STATUS_MESSAGE = 'Failed to record the error message. Please email our support team at '
        + `<a href="mailto:${SUPPORT_EMAIL}">${SUPPORT_EMAIL}</a>.`;
```

chowyb replied:
@HirdayGupta yep, seems good 👍 this would be better than the suggestion I had earlier.

HirdayGupta commented:
Done @chowyb. Ready for further review.

chowyb commented:
Looks okay now 👍 just pending adding in tests, and fixing existing ones.

HirdayGupta commented:
@chowyb Is there a way to test if something has been logged or not? Like how would I go about testing the contents of the log?

chowyb commented:
@HirdayGupta you can use `InstructorFeedbackQuestionAddActionTest.testExecuteAndPostProcessConstSumRecipient()` as a reference. More examples can be found by checking the call hierarchy of `AssertHelper.assertLogMessageEquals()`.

HirdayGupta commented:
@chowyb sorry for the delay. Been working on some other PRs. Have fixed the build here by adding in a placeholder replacement for the `Config.SUPPORT_EMAIL` as that will change across app instances. 

I was going through the existing tests and was wondering where I should place the test for this feature. One possible option would be in `AllAccessControlUiTests` under the `testPubliclyAccessiblePages()` method (where error page content is currently checked). However, since the pageobject being used there (`currentPage`) is just a parent class `AppPage` object I'd have to write the test method in the AppPage parent class. I'm not too sure how to create my own AppPage child class, but would you recommend doing so? What would you recommend to do for this test?

chowyb commented:
@HirdayGupta The `AppPage` class is intended to hold methods that can be used by multiple pages. In this case, the methods for the error page are likely to be specific to that one page only, and as such it would be better to create your own child class. You may also need to refer to `AdminHomePageUiTest.testCreateInstructorAction()` to see how getting a new AppPage class is done (in this example, it is done for the `instructorHomePage` variable.

For the tests, I was considering having an action test to check that the log message is correct when supplied with the correct parameters (similar to the example given in the previous comment), and have a UI test to check that the returned status message is correct when you click "send feedback". However, this test could be rather brittle since returning a "success" status message does not necessarily guarantee the log message was generated correctly. Changing the behaviour of "send feedback" by adding the details of the log message (subject etc.) to the returned status message could better guarantee that the log message was properly generated on the UI side, but may unnecessarily increase verbosity of the status message. I'll leave it to you to decide whether adding on to the returned status message should be done.

HirdayGupta commented:
Hi @chowyb,

**Summary of Changes**
- Refactored my new action `ErrorUserReportLogAction` to make it more testable in 5ab774a and 9db0bbc.
    - Extracted the log message generation to a method to make it easily testable
    - Made the variables that store the params private class variables to make method better testable (as now, when I call the `getUserErrorReportLogMessage()` from the test file I will also verify if the params were successfully delivered).
    - **Question: Should I add calls to `assertPostParamNotNull` in the action?**

- Added a *ActionTest to check the logging and status message returning behaviour of the action. 
    - Note that I have not used `AssertHelper.assertLogMessageEquals*()` methods here. The reason is two fold. First and foremost, that method is to check for logs in a specific format, ie, the page visit log format, but I wanted to check a log in my own format. Second, those methods are essentially glorified `assertEquals()` methods that work for a certain type of formatted string. The method checks the functioning of the `action.getLogMessage()` function rather than actually verifying whether the `log.info/log.severe, etc` methods actually write to the log (I guess/hope this is done elsewhere). Realising this, and realising that I had no need to check the message generated by `action.getLogMessage` in this test, I refactored my action as above and ran `assertEquals` with expected log message string declared by me and the log message generated by my new method.
    - Note that I have also added accounts logic in this test to get the basic `AccountsAttributes` object from the `InstructorAttributes` used to login. (If I try to use the `AccountsAttributes` object from the start by using `dataBundle.accounts.get()`, I get an `AssertionError` for my `expectedStatusMessage` because of a misisng `createdAt` field (see diff [here](https://www.diffchecker.com/FI4Gbbja)). No idea why this happens.

**Questions about the UI test:**

I had a few questions about the UI tests after going through the examples you pointed me towards.

- I intend to place the UI tests for this feature in the `AllAccessControlUiTests.testPubliclyAccessiblePages()` method. As such, I was wondering if it would be necessary to get a new AppPage instance like the `AdminHomePageUiTest.testCreateInstructorAction()` example you mentioned in your [comment above](https://github.com/TEAMMATES/teammates/pull/7675#issuecomment-316902777), or would it be ok if I simply did a `currenPage.changePageType(newPage.class)`. What is the difference betweent these two approaches and why is it better to choose one over the other?

chowyb commented:
@HirdayGupta the ActionTest looks okay 👍 adding `assertPostParamNotNull` should be fine too. Looking at the code, the reason why `dataBundle.accounts.get()` does not work is due to it retrieving the information directly from the JSON file, which does not contain a `createdAt` field. Meanwhile, the attributes themselves do have a `createdAt` field in the database after being created, resulting in a difference.

For the UI test, it seems that it may actually be preferable to use `currentPage.changePageType(newPage.class)` when looking at the number of times the method has been used. `changePageType()` may have been introduced at some point in the past to cut down on unnecessarily verbose `getNewPageInstance()` calls (although this can't really be seen via git history; we used Google Code before using GitHub). The reason I suggested the use of `getNewPageInstance()` was because I was not aware of a `changePageType()` method 😅

However, if you use this method, please note that `currentPage.changePageType(NotAuthorizedPage.class);` as seen in `AllAccessControlUiTests` is _not_ the right way to use the method: instead, the returned page from the method call should be stored as a separate variable, and that variable should be used for the new page.

HirdayGupta commented:
> Looking at the code, the reason why dataBundle.accounts.get() does not work is due to it retrieving the information directly from the JSON file, which does not contain a createdAt field. Meanwhile, the attributes themselves do have a createdAt field in the database after being created, resulting in a difference.

@chowyb The same difference remains even after I added the `createdAt` field to the JSON data. If the current code is okay, I'll stop bothering with this. Let me know :)

Thanks for the rest of the clarifications and the heads up on `changePageType`. I'll write up the UI test by tonight. :)

chowyb commented:
@HirdayGupta yep, just stick to the current one for the action test 👍

HirdayGupta commented:
@chowyb UI test added! Ready for reveiw :)

chowyb commented on src/main/webapp/dev/js/main/errorPageEmailComposer.es6:
```diff
@@ -0,0 +1,55 @@
+        const $form = $(event.target);
+        // Use Ajax to submit form data
+        $.ajax({
+            url: '/page/errorFeedbackSubmit',
```
Might be better to extract this as a constant.

chowyb commented on src/test/java/teammates/test/cases/action/ErrorUserReportLogActionTest.java:
```diff
@@ -0,0 +1,70 @@
+
+    @Override
+    protected void testAccessControl() throws Exception {
+        // TODO Auto-generated method stub
```
Might need to implement this. I'd suggest `InstructorCourseAddActionTest` as an example, but I'm not very familiar with this either.

chowyb commented on src/test/java/teammates/test/pageobjects/UserErrorReportPage.java:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    public void fillFormAndClickSubmit() {
+        fillTextBox(errorFeedbackEmailContentField, "This is an error report.");
```
May be better to supply the string as a parameter instead.

HirdayGupta commented:
@chowyb Ready for review. 

A question regarding the update to the `testAccessControl()` method: When I add in the `@Test` annotation, which I believe is what makes the method run as a test when the class is run as  TestNG Test, the test fails with throwing the exception:

```
teammates.common.exception.UnauthorizedAccessException: Unregistered user for a page that needs registration
```

I'm not sure why this happens or how to fix it. I also don't know if `verifyOnlyLoggedInUsersCanAccess` is the right method here, because technically, any user can access the form. Would `verifyAccessibleWithoutLogin` be a better choice here?

chowyb commented:
@HirdayGupta Looking at the methods in `BaseActionTest`, it seems that `verifyAnyRegisteredUserCanAccess` should be the better choice here. The access permissions here should test for who is able to submit the form, as opposed to who is able to view it. `verifyOnlyLoggedInUsersCanAccess` seems to consider that unregistered users (ie. students who submit responses via a link instead of via a student account) should be allowed to submit the form, which doesn't apply here since they have no corresponding `Account` or `AccountAttributes`.

However, I acknowledge that this is still speculation, and I could be wrong 😅

HirdayGupta commented:
@chowyb The speculation worked! 🎉 

Build should pass. Ready for review :)

damithc commented:
@HirdayGupta can you post latest screenshots?

HirdayGupta commented:
@damithc  As requested:

Screenshot of Form:

<img width="1280" alt="screen shot 2017-07-27 at 10 26 58 am" src="https://user-images.githubusercontent.com/18088721/28655057-a2d3cd5e-72b6-11e7-85f7-5658a7eac2fd.png">

**GIF of behaviour on success:**
![ezgif com-video-to-gif-2](https://user-images.githubusercontent.com/18088721/28655180-6c0ef3f6-72b7-11e7-927e-5fb8030a96e6.gif)


**GIF of behaviour on error**: (only when user is not logged in)

![ezgif com-video-to-gif-3](https://user-images.githubusercontent.com/18088721/28655201-ad9d5dbc-72b7-11e7-84b5-2957239d74ea.gif)


**Quick question:** What is the value stored in `Config.SUPPORT_EMAIL` on the production app?

HirdayGupta commented:
@damithc @chowyb Sorry for the change after approval but I added `Assumption.assertPostParamNotNull()` in 8a318cb after I realised that if someone directly accessed the URL `/page/errorFeedbackSubmit` a sever error message would be logged and the user would be able to see a JSON of the returned data.

chowyb commented:
@HirdayGupta that reminds me, there may be a need to verify if the referrer and the session token is correct for this action as well, since only logged in users should be able to submit the form. This may require (at the minimum) a change to `Const.SystemParams.PAGES_REQUIRING_ORIGIN_VALIDATION`, as well as adding a session token to the parameters that are passed in to the action. Perhaps @whipermr5 could weigh in on this as well, since he was one of the developers working on that security feature.

HirdayGupta commented:
@chowyb Additional security measures were going to be my next question 😅 

@whipermr5 Could you point me to some examples and briefly explain how exactly we use the session token?

damithc commented:
How about we change the text as follows:

># Uh oh! Something went wrong.
>
>We are sorry this happened. You can safely ignore this error page in the following cases:
>* Retrying the same action a few minutes later succeeds (i.e. no more error page).
>* You loaded an outdated page unintentionally. e.g. some browsers auto-load the pages that were loaded in the previous browsing session. 
>
>However, if you keep seeing this page on multiple failed attempts when trying to perform some action in TEAMMATES, please help us troubleshoot the problem by providing us some additional details using the form below. 

`Content` -> `More info`

HirdayGupta commented:
@damithc Do you mean removing the email address and subject fields altogether? Or adding them in after your new message?

damithc commented:
I meant replace this part:
![image](https://user-images.githubusercontent.com/1673303/28659545-78d753e6-72e2-11e7-91af-78dcc24d2cb2.png)

whipermr5 commented:
> @whipermr5 Could you point me to some examples and briefly explain how exactly we use the session token? From what I gather so far, we basically generate it on the JS side and verify it on the Java server side. I'm not entirely sure how we do this though. Do we look for an exact match after storing the token in a DB or do we look for a pattern while verifying?

The token is generated from the backend, but passed to the frontend via a cookie (which is read from the JS side). The backend verifies that it's an exact match. Simply including the new action URI inside `PAGES_REQUIRING_ORIGIN_VALIDATION` in `Const.java` will cause the verification to happen automatically. JS side, just import `makeCsrfTokenParam` from `crypto.es6` and use it to extract the token from the cookie (example: take a look at adminHome.es6).

whipermr5 commented:
You can also just include the token as a hidden input in the form from `errorPageEmailCompose.tag`; then `makeCsrfTokenParam` won't be necessary. Example: `compose.tag`.

HirdayGupta commented:
@whipermr5 After including the page in `PAGES_REQUIRING_ORIGIN_VALIDATION`, I get an invalidOrigin error with both approaches. Even after adding the following hidden input:
```HTML
<input type="hidden" name="<%=Const.ParamsNames.SESSION_TOKEN%>" value="${data.sessionToken}">
```
or after appending ${makeCsrfTokenParam()} to my action URL.

Any suggestions on avoiding the invalid origin error?

whipermr5 commented:
What error is shown on the server log?

HirdayGupta commented:
@whipermr5 Here is the log:

```
WARNING: TEAMMATESLOG|||errorFeedbackSubmit|||Servlet Action Failure|||true|||Student|||Unknown|||hirdaygupta8080|||Unknown|||<span class="text-danger">Servlet Action failure in errorFeedbackSubmit<br>class teammates.common.exception.InvalidOriginException: 
teammates.common.exception.InvalidOriginException: Invalid session token
	at teammates.ui.controller.Action.validateOriginIfRequired(Action.java:177)
	at teammates.ui.controller.Action.init(Action.java:97)
	at teammates.ui.controller.ActionFactory.getAction(ActionFactory.java:144)
	at teammates.ui.controller.ControllerServlet.doPost(ControllerServlet.java:73)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:637)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:717)
	at org.mortbay.jetty.servlet.ServletHolder.handle(ServletHolder.java:511)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1166)
	at com.google.appengine.tools.appstats.AppstatsFilter.doFilter(AppstatsFilter.java:143)
	at teammates.ui.controller.AppstatsFilter.doFilter(AppstatsFilter.java:41)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.googlecode.objectify.ObjectifyFilter.doFilter(ObjectifyFilter.java:48)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.appengine.api.socket.dev.DevSocketFilter.doFilter(DevSocketFilter.java:74)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.appengine.tools.development.ResponseRewriterFilter.doFilter(ResponseRewriterFilter.java:134)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.appengine.tools.development.HeaderVerificationFilter.doFilter(HeaderVerificationFilter.java:34)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.appengine.api.blobstore.dev.ServeBlobFilter.doFilter(ServeBlobFilter.java:63)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.utils.servlet.TransactionCleanupFilter.doFilter(TransactionCleanupFilter.java:48)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.appengine.tools.development.StaticFileFilter.doFilter(StaticFileFilter.java:122)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.appengine.tools.development.DevAppServerModulesFilter.doDirectRequest(DevAppServerModulesFilter.java:366)
	at com.google.appengine.tools.development.DevAppServerModulesFilter.doDirectModuleRequest(DevAppServerModulesFilter.java:349)
	at com.google.appengine.tools.development.DevAppServerModulesFilter.doFilter(DevAppServerModulesFilter.java:116)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at org.mortbay.jetty.servlet.ServletHandler.handle(ServletHandler.java:388)
	at org.mortbay.jetty.security.SecurityHandler.handle(SecurityHandler.java:216)
	at org.mortbay.jetty.servlet.SessionHandler.handle(SessionHandler.java:182)
	at org.mortbay.jetty.handler.ContextHandler.handle(ContextHandler.java:765)
	at org.mortbay.jetty.webapp.WebAppContext.handle(WebAppContext.java:418)
	at com.google.appengine.tools.development.DevAppEngineWebAppContext.handle(DevAppEngineWebAppContext.java:95)
	at org.mortbay.jetty.handler.HandlerWrapper.handle(HandlerWrapper.java:152)
	at com.google.appengine.tools.development.JettyContainerService$ApiProxyHandler.handle(JettyContainerService.java:508)
	at org.mortbay.jetty.handler.HandlerWrapper.handle(HandlerWrapper.java:152)
	at org.mortbay.jetty.Server.handle(Server.java:326)
	at org.mortbay.jetty.HttpConnection.handleRequest(HttpConnection.java:542)
	at org.mortbay.jetty.HttpConnection$RequestHandler.content(HttpConnection.java:938)
	at org.mortbay.jetty.HttpParser.parseNext(HttpParser.java:755)
	at org.mortbay.jetty.HttpParser.parseAvailable(HttpParser.java:218)
	at org.mortbay.jetty.HttpConnection.handle(HttpConnection.java:404)
	at org.mortbay.io.nio.SelectChannelEndPoint.run(SelectChannelEndPoint.java:409)
	at org.mortbay.thread.QueuedThreadPool$PoolThread.run(QueuedThreadPool.java:582)
<br>{
  "errorfeedbackemailreceiveraddress": [
    "app_admin@gmail.com"
  ],
  "errorfeedbackemailcontent": [
    "aa"
  ],
  "errorfeedbackemailsubject": [
    "User-submitted Error Report"
  ],
  "token": [
    "null",
    ""
  ]
}</span>|||/page/errorFeedbackSubmit?token=null|||hirdaygupta8080%20170727222123412
```

whipermr5 commented:
@HirdayGupta I realised it's because the error page is not an Action, and hence doesn't go through the normal process of providing a token (to understand it, see [these lines](https://github.com/TEAMMATES/teammates/blob/705bd0f5465ebc5351be5efdfd31dcbdd698ac3c/src/main/java/teammates/ui/controller/ControllerServlet.java#L75-L76) of `ControllerServlet` - the token is written to the page data in the try block within the Action's `executeAndPostProcess` method, and also via the cookie in `writeSessionTokenToCookieIfRequired`).

I suggest modifying the [final catch block](https://github.com/TEAMMATES/teammates/blob/705bd0f5465ebc5351be5efdfd31dcbdd698ac3c/src/main/java/teammates/ui/controller/ControllerServlet.java#L149-L156) in `ControllerServlet` to call `writeSessionTokenToCookieIfRequired` before `resp.sendRedirect(Const.ViewURIs.ERROR_PAGE)`. `writeSessionTokenToCookieIfRequired` will need to be modified to `public static void` to achieve this.

HirdayGupta commented:
@whipermr5 Since I've put this form on all error page (invalid origin, page not found, etc), wouldn't I have to place the method in all the catch blocks, not just for the generic error?

whipermr5 commented:
@HirdayGupta Oh! In that case, move the call to `writeSessionTokenToCookieIfRequired` to before the try block instead.

HirdayGupta commented:
@whipermr5 I made the `writeSessionTokenToCookieIfRequired(req, resp)` method static and added this line before the try block (And removed the line that called the method inside the try block)
```Java
ActionResult.writeSessionTokenToCookieIfRequired(req, resp);
```

I still get the same error for some reason. Requesting your assistance with this.

whipermr5 commented:
@HirdayGupta Note that this only works for `makeCsrfTokenParam`, which extracts the token from the cookie.

HirdayGupta commented:
@whipermr5 I tried it out with both, and it did not work.

whipermr5 commented:
@HirdayGupta How are you accessing the page to submit feedback? If you're accessing `*.jsp` directly (not through a redirect), it won't work because such URLs are routed directly to the relevant JSP without going through `ControllerServlet`.

whipermr5 commented:
Actually, I realised we probably don't want to put this through origin validation - even if it's set up correctly, a person might encounter an invalid origin error (due to disabling cookies etc.) and want to give feedback. That person would only encounter another invalid origin error when submitting feedback, leading to frustration.

Instead of origin validation, we might want to look into adding a captcha instead so that this feedback feature is not abused.

damithc commented:
> Actually, I realised we probably don't want to put this through origin validation - even if it's set up correctly, a person might encounter an invalid origin error (due to disabling cookies etc.) and want to give feedback. That person would only encounter another invalid origin error when submitting feedback, leading to frustration.
> 
> Instead of origin validation, we might want to look into adding a captcha instead so that this feedback feature is not abused.

Good point. But captcha introduces one more dependency on a third party system? Any alternative? What's the risk of leaving it unprotected? I don't think people can gain anything be abusing that page.

whipermr5 commented:
@damithc The only risk would be spam bots using it to email marketing materials to the admin (happens a lot for WordPress comments). I guess this is not likely to happen, and we can add a captcha next time if it does. I'm okay with leaving it unprotected for now.

damithc commented:
> I'm okay with leaving it unprotected for now.

Yes, let's do that and see how it goes.

whipermr5 commented on src/main/java/teammates/common/util/Const.java:
```diff
@@ -97,6 +97,7 @@
     public static final Date TIME_REPRESENTS_LATER;
     public static final Date TIME_REPRESENTS_NOW;
     public static final Date TIME_REPRESENTS_DEFAULT_TIMESTAMP;
+    public static final String ERROR_FEEDBACK_EMAIL_SUBJECT = "User-submitted Error Report";
```
This should be separated from the previous section by a line break.

whipermr5 commented on src/main/java/teammates/common/util/Const.java:
```diff
@@ -844,6 +845,9 @@ private Const() {
 
         public static final String SEARCH_STUDENTS = "searchstudents";
         public static final String SEARCH_COMMENTS_FOR_RESPONSES = "searchcommentforresponses";
+        public static final String ERROR_FEEDBACK_EMAIL_RECEIVER_ADDRESS = "errorfeedbackemailreceiveraddress";
```
Same here.

whipermr5 commented on src/main/java/teammates/common/util/Const.java:
```diff
@@ -1008,6 +1012,7 @@ private Const() {
         public static final String AUTOMATED_FEEDBACK_CLOSED_REMINDERS = "/auto/feedbackSessionClosedReminders";
         public static final String AUTOMATED_FEEDBACK_CLOSING_REMINDERS = "/auto/feedbackSessionClosingReminders";
         public static final String AUTOMATED_FEEDBACK_PUBLISHED_REMINDERS = "/auto/feedbackSessionPublishedReminders";
+        public static final String ERROR_FEEDBACK_SUBMIT = "/page/errorFeedbackSubmit";
```
Same here.

whipermr5 commented on src/main/java/teammates/common/util/Const.java:
```diff
@@ -1468,6 +1473,8 @@ private Const() {
         public static final String ADMIN_LOG_INSTRUCTOR_COURSE_ENROLL_PAGE_LOAD =
                 "instructorCourseEnroll Page Load" + Const.HTML_BR_TAG
                 + "Enrollment for Course <span class=\"bold\">[%s]</span>";
+        public static final String ERROR_FEEDBACK_SUBMIT_SUCCESS = "Your error report has been recorded. "
```
Same here.

whipermr5 commented on src/main/java/teammates/ui/controller/ActionFactory.java:
```diff
@@ -125,6 +125,7 @@
 
         map(CREATE_IMAGE_UPLOAD_URL, CreateImageUploadUrlAction.class);
         map(IMAGE_UPLOAD, ImageUploadAction.class);
+        map(ERROR_FEEDBACK_SUBMIT, ErrorUserReportLogAction.class);
```
Same here.

whipermr5 commented on src/main/webapp/deadlineExceededErrorPage.jsp:
```diff
@@ -1,6 +1,7 @@
 <%@ taglib tagdir="/WEB-INF/tags" prefix="t" %>
 <% response.setStatus(500);%>
 <t:errorPage>
+  <t:statusMessage statusMessagesToUser="${data.statusMessagesToUser}" />
```
I believe this can be placed in `errorPage.tag` before `doBody` to avoid repeating it across all the error page `jsp`s.

HirdayGupta commented:
> The only risk would be spam bots using it to email marketing materials to the admin (happens a lot for WordPress comments).

Also, don't forget, the user has to be logged in to successfully submit feedback. So it would have to be one of our own users who would have to use these bots to be successful. (Which I think makes it even more unlikely, unless some of the students decide to mess around :p)

Just for reference: **Changes made after @chowyb had approved**

- Add `Assumption.assertPostParamNotNull()` in 8a318cb
- Rephrase the error message as per @damithc's suggestion in https://github.com/TEAMMATES/teammates/pull/7675#issuecomment-318265630 in commit e205a3e
- Update tests for the changes made above in 6bf1d11 and 52d4ba5

**New GIFs:**

Success:

![4315-success](http://g.recordit.co/21NFH3tk3r.gif)

Error:

![4315-error](http://g.recordit.co/8A2vvAoYAR.gif)

damithc commented:
> Also, don't forget, the user has to be logged in to successfully submit feedback.

We allow submitting responses without login. 🤔

HirdayGupta commented:
> We allow submitting responses without login. 

@damithc No sir, I had mentioned it earlier and asked if reviewers wanted it changed. In fact, the only time that error message in recording the feedback is displayed is when the user is not logged in.

whipermr5 commented:
@HirdayGupta He means that we should allow students to submit error feedback even when they are accessing feedback response submission pages via unique link (i.e., not logged in).

whipermr5 commented:
@HirdayGupta In that case, the registration key is used for authentication. Adding your new action URI to `PAGES_ACCESSIBLE_WITHOUT_GOOGLE_LOGIN` should suffice. Also update the tests.

HirdayGupta commented:
> He means that we should allow students to submit error feedback even when they are accessing feedback response submission pages via unique link (i.e., not logged in).

Ahh, my bad 😅 

> In that case, the registration key is used for authentication. Adding your new action URI to PAGES_ACCESSIBLE_WITHOUT_GOOGLE_LOGIN should suffice

Just tried this out and it still sends a redirect to the login page for some reason. Not sure why. I believe that doPost in ControllerServlet is responsible for this redirect, and I can't seem to find where it checks the `PAGES_ACCESSIBLE_WITHOUT_GOOGLE_LOGIN` before doing so. Requesting help with this.

> Also update the tests.

They don't need updating with the recent changes. HTML doesn't change.

whipermr5 commented:
@HirdayGupta I just got it to submit successfully by manually including `key`, `courseid` and `studentemail` in the POST params (using hidden inputs in the form). These parameters are included in the unique link, and are required for the registration key authentication to work. We need to find a way to pass these parameters to the error page when, say, the user tries to submit feedback and an error occurs.

whipermr5 commented:
> We need to find a way to pass these parameters to the error page when, say, the user tries to submit feedback and an error occurs.

This can be done by appending those parameters to the URL before we `resp.sendRedirect` in each of the catch blocks in `ControllerServlet`. In addition to these parameters, we can also pass in the requested URL that resulted in the error and include that in the error report, which would help us in debugging.

HirdayGupta commented:
@whipermr5 is it necessary to pass these parameters to the error pages each time? Technically, we only should be putting these params if a student who has not yet registered is trying to access the error pages, right?

Either way, I'm running into some trouble accessing the required parameters (courseId, key, email) in the catch blocks. Most of the information I need seems to be stored in the `c` Action object in the ControllerServlet's doPost method's try block. Any suggestions on how I can access this object in the catch blocks? I tried declaring it outside the try block but got `variable may not have been initialized` compilation errors.

whipermr5 commented:
@HirdayGupta The information should be in the parameters. Yes; they only need to be passed if they are present in the existing parameters.

HirdayGupta commented:
@whipermr5 added code that implements the mechanism you suggested in https://github.com/TEAMMATES/teammates/pull/7675#issuecomment-319304601. Wrote what I thought was best. Submitting error reports for unregistered students who are submitting using the link works with the most recent changes. I don't know if my implementation is the best one, but it was my best shot. **Ready for review :)**

whipermr5 commented on src/main/java/teammates/ui/controller/ControllerServlet.java:
```diff
@@ -153,12 +153,25 @@ public final void doPost(HttpServletRequest req, HttpServletResponse resp) throw
+            redirectUrl.append('?')
+                       .append(Const.ParamsNames.REGKEY + '=' + params.get(Const.ParamsNames.REGKEY)[0] + '&')
+                       .append(Const.ParamsNames.COURSE_ID + '=' + params.get(Const.ParamsNames.COURSE_ID)[0] + '&')
+                       .append(Const.ParamsNames.STUDENT_EMAIL + '=' + params.get(Const.ParamsNames.STUDENT_EMAIL)[0]);
```
We have a helper method `addParamToUrl` written for this job. It also handles null values by skipping them, so the null checks above aren't needed.

HirdayGupta replied:
@whipermr5 Just tried this out using the addParamToUrl method. The problem is that `params.get("key")` returns a `String[]` array object not just a string, which means I have to reference the 0th element like I have done above to get the value required. Now, without the null checks above, this results in a nullpointer exception. Should I just keep the null checks and replace the StringBuffer appending with the `addParamToUrl` helper method or is there a better way?

whipermr5 replied:
I just took a look at `HttpRequestHelper` again and noticed a method `getValueFromParamMap`.

HirdayGupta commented:
@whipermr5 Thanks for the suggestions! Ready for review. :)

HirdayGupta commented:
**Summary of changes from last 2 commits:**

Travis pointed out some tests that were failing because of the new feature introduced in https://github.com/TEAMMATES/teammates/pull/7675/commits/3c0a7821a1d1492a2a43df03d06b69b903fec9a3. Fixed these tests using the approaches described below:
- **InstructorFeedbackEditPageUiTest**: Changed the `assertEquals(expected, actual)` to `assertTrue(actual.contains(expected)`. I feel that this approach is a good balance between accuracy and complexity. Since I also added the `requestedurl` param, this assertion now tests the essence of the expected URL, without the added complexity of going into too many specifics and adding too many more parameters.

- **InstructorHomePageUiTest:** Changed the `endsWith` to `contains` for the same reason described above. Also added the `requestedurl` param in the tested string.

**Ready for review**

HirdayGupta commented:
@whipermr5 Reminder :)

HirdayGupta commented:
@whipermr5 Please review

HirdayGupta commented:
>  @whipermr5 there's a review request from you as well, probably due to the involvement of session tokens.

Actually, @damithc if you read [this comment](https://github.com/TEAMMATES/teammates/pull/7675#issuecomment-319253818) and a few below it, you'll see that we decided to drop the session tokens idea as it seemed unnecessary for this page.

The review request from @whipermr5 is because he had reviewed some parts of this PR after @chowyb and because he helped in adding support for unregistered students to submit feedback (even though they aren't technically logged in). It is just this one last aspect of the PR (from https://github.com/TEAMMATES/teammates/pull/7675/commits/d352e33a6e751e4b4feb8399f3eb6e3f3f6cdb93 onwards) that remains unreviewed.

whipermr5 commented on src/main/webapp/WEB-INF/tags/errorPage.tag:
```diff
@@ -26,8 +26,36 @@
     </div>
     <t:bodyFooter />
 </body>
+<script type="text/javascript" src="https://unpkg.com/jquery@1.12.4/dist/jquery.min.js"></script>
```
This should not be hardcoded - use `FrontEndLibrary`.

HirdayGupta commented:
@whipermr5 made the change. Ready for review!

HirdayGupta commented:
@whipermr5 Please review this.

whipermr5 commented:
@HirdayGupta Oops, there's a conflict now (use 2-space indentation for jsp and tag files).

HirdayGupta commented:
@whipermr5 Resolved the conflicts. Ready for merge!

whipermr5 commented:
@HirdayGupta 👍 Will merge this next!

whipermr5 commented on src/main/webapp/WEB-INF/tags/errorPageEmailCompose.tag:
```diff
@@ -0,0 +1,25 @@
+<%@ tag import="teammates.common.util.Config" %>
+<%@ tag import="teammates.common.util.Const" %>
+<form action="<%= Const.ActionURIs.ERROR_FEEDBACK_SUBMIT %>" method="post" id="error-feedback-form">
+    <div class="form-group">
```
Sorry just noticed; this file needs to use 2-space indentation too.

HirdayGupta replied:
Good catch again! On it. 😂

HirdayGupta commented:
@whipermr5 Done! 2-space indented now. Good to go (I hope) 😅

HirdayGupta commented:
@damithc Sir, please do let me know here, or anywhere, how this feature works out for the App and if it comes in handy. And also let me know of any new functionality you would like added that might make this more useful as and when the user error reports come in - no matter even if it is after the GSoC coding period, I intend to keep contributing. :)

damithc commented:
> @damithc Sir, please do let me know here, or anywhere, how this feature works out for the App and if it comes in handy. And also let me know of any new functionality you would like added that might make this more useful as and when the user error reports come in - no matter even if it is after the GSoC coding period, I intend to keep contributing. :)

👍 I'll let you know when users start using this feature.
